### PR TITLE
ADBDEV-6797: Make the wasTerminated variable atomic

### PR DESCRIPTION
**Pull request by RekGRpth**
Make the wasTerminated variable atomic

The global variable wasTerminated must be atomic because it is used
simultaneously in several different goroutines.

No tests were added because it is a rare case of a race condition that is
nearly impossible to reproduce in a stable manner.

### CODE REVIEW

File: backup/backup.go [hunk 1]
```diff
@@ -314,7 +314,7 @@ func backupData(tables []Table) {
 		initialPipes := CreateInitialSegmentPipes(oidList, globalCluster, connectionPool, globalFPInfo)
 		// Do not pass through the --on-error-continue flag or the resizeClusterMap because neither apply to gpbackup
 		utils.StartGpbackupHelpers(globalCluster, globalFPInfo, "--backup-agent",
-			MustGetFlagString(options.PLUGIN_CONFIG), compressStr, false, false, &wasTerminated, initialPipes, true, false, 0, 0, gplog.GetVerbosity())
+			MustGetFlagString(options.PLUGIN_CONFIG), compressStr, false, false, wasTerminated.Load(), initialPipes, true, false, 0, 0, gplog.GetVerbosity())
```
Comment (dkovalev1) [inline backup/backup.go:317]: It seems to be not the best idea to pass a copy of an atomic value to the function. And semantically it differs from existing version. I suggest to pass a pointer to wasTerminated here, as before.

File: backup/global_variables.go [hunk 2]
```diff
@@ -27,7 +28,7 @@ const (
 	Deferred
 	Complete
 	PG_LOCK_NOT_AVAILABLE = "55P03"
-	ENUM_TYPE_OID = 3500
+	ENUM_TYPE_OID         = 3500
```
Comment (dkovalev1) [inline backup/global_variables.go:31]: it's not relevant to this commit

File: helper/helper.go [hunk 3]
```diff
@@ -138,33 +139,31 @@ func InitializeGlobals() {
...
 				}
 			case unix.SIGUSR1:
 				gplog.Warn("Received shutdown request on segment %d: beginning cleanup", *content)
-				terminatedChan <- true
+				wasTerminated.Store(true)
 			}
 		}()
-		wasTerminated = <-terminatedChan
-		if wasTerminated {
+		if wasTerminated.Load() {
```
Comment (dkovalev1) [inline helper/helper.go:166]: Wait on a channel is lost here and it looks like busy loop now. I suggest revert to channel here and store value read from the channel.

File: restore/data.go [hunk 4]
```diff
@@ -245,7 +245,7 @@ func restoreDataFromTimestamp(fpInfo filepath.FilePathInfo, dataEntries []toc.Co
 		if backupConfig.Compressed {
 			compressStr = fmt.Sprintf(" --compression-type %s ", utils.GetPipeThroughProgram().Name)
 		}
-		utils.StartGpbackupHelpers(globalCluster, fpInfo, "--restore-agent", MustGetFlagString(options.PLUGIN_CONFIG), compressStr, MustGetFlagBool(options.ON_ERROR_CONTINUE), isFilter, &wasTerminated, initialPipes, backupConfig.SingleDataFile, resizeCluster, origSize, destSize, gplog.GetVerbosity())
+		utils.StartGpbackupHelpers(globalCluster, fpInfo, "--restore-agent", MustGetFlagString(options.PLUGIN_CONFIG), compressStr, MustGetFlagBool(options.ON_ERROR_CONTINUE), isFilter, wasTerminated.Load(), initialPipes, backupConfig.SingleDataFile, resizeCluster, origSize, destSize, gplog.GetVerbosity())
```
Comment (dkovalev1) [inline restore/data.go:248]: Same as above, IMHO it's better to pass a pointer to atomic here.

File: utils/agent_remote.go [hunk 5]
```diff
@@ -144,12 +144,12 @@ func VerifyHelperVersionOnSegments(version string, c *cluster.Cluster) {
...
 }
 
-func StartGpbackupHelpers(c *cluster.Cluster, fpInfo filepath.FilePathInfo, operation string, pluginConfigFile string, compressStr string, onErrorContinue bool, isFilter bool, wasTerminated *bool, copyQueue int, isSingleDataFile bool, resizeCluster bool, origSize int, destSize int, verbosity int) {
+func StartGpbackupHelpers(c *cluster.Cluster, fpInfo filepath.FilePathInfo, operation string, pluginConfigFile string, compressStr string, onErrorContinue bool, isFilter bool, wasTerminated bool, copyQueue int, isSingleDataFile bool, resizeCluster bool, origSize int, destSize int, verbosity int) {
 	// A mutex lock for cleaning up and starting gpbackup helpers prevents a
 	// race condition that causes gpbackup_helpers to be orphaned if
 	// gpbackup_helper cleanup happens before they are started.
 	helperMutex.Lock()
-	if *wasTerminated {
+	if wasTerminated {
```
Comment (dkovalev1) [inline utils/agent_remote.go:152]: Here we check a copy of variable which could be changed from it's Load()

File: backup/backup.go (hunk 1 above)
Comment (RekGRpth) [reply backup/backup.go]: done

File: backup/global_variables.go (hunk 2 above)
Comment (RekGRpth) [reply backup/global_variables.go:31]: This was done by the default formatter in the VScode when saving the file.

File: helper/helper.go (hunk 3 above)
Comment (RekGRpth) [reply helper/helper.go:168]: done

File: restore/data.go (hunk 4 above)
Comment (RekGRpth) [reply restore/data.go]: done

File: utils/agent_remote.go (hunk 5 above)
Comment (RekGRpth) [reply utils/agent_remote.go]: fixed

**Comment by dkovalev1** [comment]
May be it worst mentioning that this is rare case for race condition, so it's almost impossible to reproduce in stable way, that's why there are no tests for it ?

**Comment by RekGRpth** [comment]
> May be it worst mentioning that this is rare case for race condition, so it's almost impossible to reproduce in stable way, that's why there are no tests for it ?

added

**Comment by dkovalev1** [comment]
> > May be it worst mentioning that this is rare case for race condition, so it's almost impossible to reproduce in stable way, that's why there are no tests for it ?
> 
> added

great

**Review by dkovalev1 — APPROVED** [review]

**Review by KnightMurloc — APPROVED** [review]

**Pull request merged**